drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1570 @@
+
+	foo_data.socket_data[id].connected = 1;
+}
+
+static void on_cmd_conn_closed(struct net_buf **buf, u16_t len)
+{
+	int id;
+	u8_t msg[2];
+	struct socket_data *sock = NULL;
+
+	net_buf_linearize(msg, 1, *buf, 0, 1);
+	msg[1] = '\0';
+	id = atoi(msg);
+
+	sock = &foo_data.socket_data[id];
+	sock->connected = 0;
+}
+
+static void on_cmd_sent_bytes(struct net_buf **buf, u16_t len)
+{
+	char temp[32];
+	char *tok;
+
+	net_buf_linearize(temp, 32, *buf, 0, len);
+	temp[len] = '\0';
+
+	tok = strtok(temp, " ");
+	while(tok) {
+		tok = strtok(NULL, " ");
+	}
+}
+
+
+static void on_cmd_wifi_scan_resp(struct net_buf **buf, u16_t len)
+{
+	int i;
+	char temp[32];
+	struct wifi_scan_result result;
+	int delimiters[6];
+	int slen;
+
+	/* ecn, ssid, rssi, mac, channel, freq */
+
+	delimiters[0] = 1;
+	for (i = 1; i < 6; i++) {
+		delimiters[i] = net_buf_find_next_delimiter(*buf, ',',
+					delimiters[i-1] + 1);
+		if (delimiters[i] == -1) {
+			return;
+		}
+		delimiters[i]++;
+	}
+
+	/* ecn */
+	net_buf_linearize(temp, 1, *buf, delimiters[0],
+			  delimiters[1] - delimiters[0]);
+	if (temp[0] != '0') {
+		result.security = WIFI_SECURITY_TYPE_PSK;
+	} else {
+		result.security = WIFI_SECURITY_TYPE_PSK;
+	}
+
+	/* ssid */
+	slen = delimiters[2] - delimiters[1] - 3;
+	net_buf_linearize(result.ssid, 32, *buf, delimiters[1] + 1,
+			  slen);
+	result.ssid_length = slen;
+
+	/* rssi */
+	slen = delimiters[3] - delimiters[2];
+	net_buf_linearize(temp, 32, *buf, delimiters[2], slen);
+	temp[slen] = '\0';
+	result.rssi = strtol(temp, NULL, 10);
+
+	/* channel */
+	slen = delimiters[5] - delimiters[4];
+	net_buf_linearize(temp, 32, *buf, delimiters[4], slen);
+	temp[slen] = '\0';
+	result.channel = strtol(temp, NULL, 10);
+
+	/* issue callback to report scan results */
+	if (foo_data.wifi_scan_cb) {
+		foo_data.wifi_scan_cb(foo_data.iface, 0, &result);
+	}
+}
+
+static const char nm_label[] = "netmask";
+static const char gw_label[] = "gateway";
+static const char ip_label[] = "ip";
+
+static void on_cmd_ip_addr_resp(struct net_buf **buf, u16_t len)
+{
+	char ip_addr[] = "xxx.xxx.xxx.xxx";
+	int d[3];
+	int slen;
+
+	d[0] = net_buf_find_next_delimiter(*buf, ':', 0);
+	d[1] = net_buf_find_next_delimiter(*buf, '\"', d[0] + 1);
+	d[2] = net_buf_find_next_delimiter(*buf, '\"', d[1] + 1);
+
+	slen = d[2] - d[1] - 1;
+
+	net_buf_linearize(ip_addr, 16, *buf, d[0] + 2, slen);
+	ip_addr[slen] = '\0';
+
+	if (net_buf_ncmp(*buf, nm_label, strlen(nm_label)) == 0) {
+		inet_pton(AF_INET, ip_addr,
+			  &foo_data.netmask);
+	} else if (net_buf_ncmp(*buf, ip_label, strlen(ip_label)) == 0) {
+		inet_pton(AF_INET, ip_addr,
+			  &foo_data.ip);
+	} else if (net_buf_ncmp(*buf, gw_label, strlen(gw_label)) == 0) {
+		inet_pton(AF_INET, ip_addr,
+			  &foo_data.gw);
+	} else {
+		return;
+	}
+}
+
+static void on_cmd_mac_addr_resp(struct net_buf **buf, u16_t len)
+{
+	int i;
+	char octet[] = "xx";
+	for (i = 0; i < 6; i++) {
+		net_buf_pull_u8(*buf);
+		octet[0]= net_buf_pull_u8(*buf);
```

Comment:
Note that net_buf_pull*() works only within one fragment, if your data could expand to multiple buffers, then you should use net_frag_read_u8() instead.